provider/pkg/gen/discovery.go

```diff
@@ -0,0 +1,133 @@
+// Copyright 2016-2021, Pulumi Corporation.
+
+package gen
+
+import (
+	"fmt"
+	"github.com/gedex/inflector"
+	"github.com/pkg/errors"
+	"github.com/pulumi/pulumi/pkg/v3/codegen"
+	"google.golang.org/api/discovery/v1"
+	"strings"
+)
+
+// discoveryDocumentResource is a combination of REST methods that should be treated as
+// a fully capable Pulumi resource for the generation purpose.
+type discoveryDocumentResource struct {
+	createMethod, getMethod, updateMethod, deleteMethod *discovery.RestMethod
+}
+
+// findResources builds a map of resources for a given Discovery Document resource. Map keys are resource names and
+// values are pointers to CRUD REST methods.
+func findResources(docName string, rest map[string]discovery.RestResource) (map[string]discoveryDocumentResource, error) {
+	result := map[string]discoveryDocumentResource{}
+	add := func(typeName string, dd discoveryDocumentResource) {
+		addFoundResource(result, typeName, dd)
+	}
+	err := findResourcesImpl(docName, "", rest, add)
+	if err != nil {
+		return nil, err
+	}
+	return result, nil
+}
+
+func findResourcesImpl(docName, parentName string, rest map[string]discovery.RestResource,
+	add func(string, discoveryDocumentResource)) error {
+	var postMethods []discovery.RestMethod
+	for _, resourceName := range codegen.SortedKeys(rest) {
+		res := rest[resourceName]
+		name := ToUpperCamel(inflector.Singularize(resourceName))
+		var createMethod, getMethod, listMethod, updateMethod, deleteMethod *discovery.RestMethod
+		for methodName, value := range res.Methods {
+			restMethod := value
+			if restMethod.HttpMethod == "POST" {
+				postMethods = append(postMethods, restMethod)
+			}
+			switch methodName {
+			case "create", "insert":
+				createMethod = &restMethod
+			case "submit", "register":
+				if createMethod == nil {
+					createMethod = &restMethod
+				}
+			case "update", "replaceService" /*special case for Cloud Run*/ :
+				updateMethod = &restMethod
+			case "patch":
+				if updateMethod == nil {
+					updateMethod = &restMethod
+				}
+			case "get":
+				getMethod = &restMethod
+			case "read":
+				if getMethod == nil {
+					getMethod = &restMethod
+				}
+			case "list":
+				listMethod = &restMethod
+			case "setIamPolicy":
+				typeName := fmt.Sprintf("%s%sIamPolicy", parentName, name)
+				if getIamPolicy, has := res.Methods["getIamPolicy"]; has {
+					dd := discoveryDocumentResource{
+						createMethod: &restMethod,
+						getMethod: &getIamPolicy,
+						updateMethod: &restMethod,
+					}
+					add(typeName, dd)
+				}
+			case "delete", "dropDatabase" /*special case for Spanner Database*/ :
+				deleteMethod = &restMethod
+			}
+		}
+
+		typeName := fmt.Sprintf("%s%s", parentName, name)
+		switch {
+		case createMethod != nil && getMethod != nil && getMethod.HttpMethod != "GET":
+			return errors.Errorf("get method %q is not supported: %s (%s)", getMethod.HttpMethod, typeName, docName)
+		case createMethod != nil && getMethod != nil:
```

Comment:
I think it might be clearer to combine these and include the `getMethod.HttpMethod != "GET"` as an if statement inside that case.

Reply:
I was shooting for the least nesting. I appreciate the feedback - will adjust the code.